rawfile/declarative.py

```diff
@@ -31,39 +32,58 @@ def be_symlink(path, to):
     path.symlink_to(to)
 
 
-def be_mounted(dev, mountpoint, readonly=False):
-    dev = Path(dev).resolve()
-    mountpoint = Path(mountpoint)
+def mount(device, mountpoint, readonly=False):
+    current_dev = get_device_for_mountpoint(mountpoint)
+    device = Path(device).resolve().as_posix()
 
-    if mountpoint.is_mount():
-        if True:  # TODO: verify that the right device is mounted
-            return
-        # noinspection PyUnreachableCode
-        be_unmounted(mountpoint)
-    elif mountpoint.resolve().is_file():
+    if current_dev:
+        current_dev = Path(current_dev).resolve().as_posix()
+        if current_dev == device:
+            return  # TODO: LOG
+        unmount(mountpoint, clear_mountpoint=False)
+
+    if Path(mountpoint).resolve().is_file():
         opts = ["bind"]
-        # todo: this is not sufficient, we need to create a separate loop device for RO
+        # TODO: this is not sufficient,
+        # we need to create a separate loop device for RO
         if readonly:
             opts.append("ro")
         opts_str = ",".join(opts)
-        run(f"mount -t none -o {opts_str} {dev} {mountpoint}")
+        run(f"mount -t none -o {opts_str} {device} {mountpoint}")
         return
-
-    fs = current_fs(dev)
-    if fs == "ext4":
-        run(f"mount -t ext4 {dev} {mountpoint}")
-    elif fs == "btrfs":
-        run(f"mount -t btrfs -o flushoncommit {dev} {mountpoint}")
-    elif fs == "xfs":
-        run(f"mount -t xfs {dev} {mountpoint}")
-    else:
-        raise Exception(f"Unsupported fs type: {fs}")
-
-
-def be_unmounted(path):
-    path = Path(path)
-    while path.is_mount():
-        run(f"umount {path}")
+    fs = from_device(device)
+    if not fs:
+        raise UnknownFileSystemError(device=device)
+    fs.mountpoint = mountpoint
+    options = []
+    # FIXME: RO mount for filesystems ignored (https://github.com/openebs/rawfile-localpv/pull/86#issuecomment-3000864823)
+    # if readonly:
+    #     options.extend(["-o", "ro"])
+    fs.mount(options=options)
+
+
+def unmount(mountpoint: str, clear_mountpoint=True):
+    path = Path(mountpoint)
+    if not path.exists():
+        return
+    if path.is_file() or path.is_block_device():
+        subprocess.run(
```

Comment:
should we use the utils "run" methods as used elsewhere in the code?
would allow us to add trace logging in a common place for verbose logging

Reply:
@tiagolobocastro
The utilities are refactored now, and I have moved all of the `run` calls to use a single function